.buildkite/scripts/build_and_upload_packages.sh

```diff
@@ -0,0 +1,94 @@
+#!/usr/bin/env bash
+
+set -euo pipefail
+
+################################################################################
+# This file is responsible for building and uploading non-Docker packages, like
+# the Firecracker microVM kernel and the Firecracker rootfs.
+# It will not upload anything if a file of that name and SHA already exist
+# upstream.
+################################################################################
+
+source .buildkite/scripts/lib/artifacts.sh
+source .buildkite/scripts/lib/version.sh
+
+TAG="$(timestamp_and_sha_version)"
+
+(
+    echo "--- Building packages"
+    make dist/firecracker_kernel.tar.gz
+)
+
+# These will be uploaded to Cludsmith with just their basename.
+readonly PACKAGES=(
+    dist/firecracker_kernel.tar.gz
+)
+
+# This is where our images will ultimately be promoted to. It is the
+# registry we'll need to query to see if file with the same contents
+# already exists.
+readonly UPSTREAM_REGISTRY="${UPSTREAM_REGISTRY:-grapl/testing}"
+readonly UPLOAD_TO_REGISTRY="${UPLOAD_TO_REGISTRY:-grapl/raw}"
+
+# This is the list of packages that actually have different shas
+new_packages=()
+
+# Returns a string like `2de4f0c...`
+sha256_of_file() {
+    sha256sum "${1}" | awk '{print $1;}'
+}
+
+# Get the json query result of a named package.
+# Usage:
+#  query_package firecracker_kernel.tar.gz
+#  CHECKSUM=a1b2c3 query_package firecracker_kernel.tar.gz
+cloudsmith_query_package() {
+    queries=(
+        --query "name:${1}"
+    )
+    if [ -n "${CHECKSUM}" ]; then
+        queries+=(
+            --query "checksum:${CHECKSUM}"
+        )
+    fi
+    cloudsmith ls packages \
+        "${UPSTREAM_REGISTRY}" \
+        "${queries[@]}" \
+        --output-format=pretty_json |
+        jq ".data"
+}
+
+# Returns 0 if it is present; 1 if not.
+present_upstream() {
+    CHECKSUM="${2}" cloudsmith_query_package "${1}" |
+        jq -e ". | length != 0"
+}
+
+echo "--- :cloudsmith::sleuth_or_spy: Checking upstream repository to determine what to promote"
+
+for artifact_path in "${PACKAGES[@]}"; do
+    artifact_name=$(basename "${artifact_path}")
+    echo "--- :cloudsmith: Checking '${artifact_name}' in '${UPSTREAM_REGISTRY}'"
+    incoming_sha256="$(sha256_of_file "${artifact_path}")"
+
+    echo "Checking if a version with SHA ${incoming_sha256} exists upstream..."
+    if ! present_upstream "${artifact_name}" "${incoming_sha256}"; then
+        echo "Package not present upstream; will promote '${artifact_name}'"
+        new_packages+=("${artifact_path}")
+    else
+        echo "Package with this SHA exists upstream; no change needed"
+    fi
+done
+
+echo "--- :cloudsmith::up: Uploading new packages to Cloudsmith"
+for artifact_path in "${new_packages[@]}"; do
+    artifact_name=$(basename "${artifact_path}")
+    cloudsmith upload raw "${UPLOAD_TO_REGISTRY}" \
+        "${artifact_path}" \
+        --name "${artifact_name}" \
+        --version "${TAG}"
```

Comment:
🤔 Does it make sense to give the Firecracker kernel image a version based on information derived from the Grapl repo? It's version `1.0.0`, not `20220215blahblahblah`

Reply:
Incorporating the actual kernel version might be a good idea, too.

Reply:
I'd argue that this captures _the state/sha of the grapl repo that generated said kernel_, and that said information can be extrapolated from it

but i don't care that much :)

Reply:
Yeah, that's true.

I'd counter-argue that what's here in the Grapl repo isn't really doing anything to influence the resulting kernel image (e.g., we're not using some custom config to build it). We're literally pulling from a tagged release of the Firecracker repo and running _its_ code.

Until we're actually doing something distinct, adding an additional layer of indirection to figure out what's actually in a given artifact doesn't seem very useful.